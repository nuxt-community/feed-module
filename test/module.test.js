jest.setTimeout(60000)

const { resolve, join } = require('path')
const { existsSync, removeSync, readFileSync } = require('fs-extra')
const { Nuxt, Builder, Generator } = require('nuxt-edge')
const request = require('request-promise-native')
const getPort = require('get-port')
const logger = require('@/logger')
const { createFeed, feedOptions } = require('./feed-options')

const config = require('./fixture/nuxt.config')
config.dev = false

let nuxt, port

logger.mockTypes(() => jest.fn())

const url = path => `http://localhost:${port}${path}`
const get = path => request(url(path))

const setupNuxt = async (config) => {
  const nuxt = new Nuxt(config)
  await nuxt.ready()
  await new Builder(nuxt).build()
  port = await getPort()
  await nuxt.listen(port)

  return nuxt
}

describe('module', () => {
  beforeEach(() => {
    logger.clear()
  })

  afterEach(async () => {
    const filePath = resolve(nuxt.options.srcDir, join(nuxt.options.dir.static, 'feed.xml'))
    if (existsSync(filePath)) {
      removeSync(filePath)
    }

    if (nuxt) {
      await nuxt.close()
    }
  })

  test('generate simple rss', async () => {
    nuxt = new Nuxt(config)
    await nuxt.ready()

    const builder = new Builder(nuxt)
    const generator = new Generator(nuxt, builder)
    await generator.generate()

    const filePath = resolve(nuxt.options.srcDir, join(nuxt.options.dir.static, 'feed.xml'))
    expect(readFileSync(filePath, { encoding: 'utf8' })).toMatchSnapshot()
  })

  test('simple rss', async () => {
    nuxt = await setupNuxt(config)

    const html = await get('/feed.xml')
    expect(html).toMatchSnapshot()
  })

  test('simple atom', async () => {
    nuxt = await setupNuxt({
      ...config,
      feed: [
        createFeed('atom1')
      ]
    })

    const html = await get('/feed.xml')
    expect(html).toMatchSnapshot()
  })

  test('simple json', async () => {
    nuxt = await setupNuxt({
      ...config,
      feed: [
        createFeed('json1')
      ]
    })

    const html = await get('/feed.xml')
    expect(html).toMatchSnapshot()
  })

  test('non-latin rss', async () => {
    nuxt = await setupNuxt({
      ...config,
      feed: [
        {
          create(feed) {
            feed.options = {
              title: 'Популярные новости России и мира',
              link: 'http://site.ru/feed.xml',
              description: 'Новости России и мира на сайте site.ru',
              updated: new Date(Date.UTC(2000, 6, 14))
            }
            feed.addContributor({
              name: 'Команда проекта site.ru',
              email: 'support@ site.ru',
              link: 'http://site.ru/'
            })
          },
          type: 'rss2'
        }
      ]
    })

    const html = await get('/feed.xml')
    expect(html).toMatchSnapshot()
  })

  test('object rss', async () => {
    nuxt = await setupNuxt({
      ...config,
      feed: {
        data: { title: 'Feed Title' },
        create(feed, data) {
          feedOptions.title = data.title
          feed.options = feedOptions
        },
        type: 'rss2'
      }
    })

    const html = await get('/feed.xml')
    expect(html).toMatchSnapshot()
  })

  test('factory rss', async () => {
    nuxt = await setupNuxt({
      ...config,
      feed: {
        data: { title: 'Feed Title' },
        factory: data => ({
          data,
          create(feed, { title }) {
            feedOptions.title = data.title
            feed.options = feedOptions
          },
          type: 'rss2'
        })
      }
    })

    const html = await get('/feed.xml')
    expect(html).toMatchSnapshot()
  })

  test('function rss', async () => {
    nuxt = await setupNuxt({
      ...config,
      feed: () => createFeed()
    })

    const html = await get('/feed.xml')
    expect(html).toMatchSnapshot()
  })

  test('multi rss', async () => {
    nuxt = await setupNuxt({
      ...config,
      feed: [
        { ...createFeed(), ...{ path: '/feed1.xml' } },
        { ...createFeed(), ...{ path: '/feed2.xml' } }
      ]
    })

    const html1 = await get('/feed1.xml')
    expect(html1).toMatchSnapshot()

    const html2 = await get('/feed2.xml')
    expect(html2).toMatchSnapshot()
  })

  test('no type set', async () => {
    nuxt = await setupNuxt({
      ...config,
      feed: [
        {}
      ]
    })

    const html = await get('/feed.xml')
    expect(html).toMatchSnapshot()

    expect(logger.fatal).toHaveBeenCalledWith('Could not create Feed /feed.xml - Unknown feed type')
  })

  test('error on create feed', async () => {
    nuxt = await setupNuxt({
      ...config,
      feed: [
        {
          create(feed) {
            throw new Error('Error on create feed')
          },
          type: 'rss2'
        }
      ]
    })

    const html = await get('/feed.xml')
    expect(html).toMatchSnapshot()

    expect(logger.error).toHaveBeenCalledWith(new Error('Error on create feed'))
    expect(logger.fatal).toHaveBeenCalledWith('Error while executing feed creation function')
  })
})
