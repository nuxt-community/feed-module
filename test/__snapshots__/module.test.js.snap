// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`module error on create feed 1`] = `""`;

exports[`module factory rss 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<rss version=\\"2.0\\" xmlns:atom=\\"http://www.w3.org/2005/Atom\\">
    <channel>
        <title>Feed Title</title>
        <link>http://example.com/</link>
        <description>This is my personal feed!</description>
        <lastBuildDate>Fri, 14 Jul 2000 00:00:00 GMT</lastBuildDate>
        <docs>https://validator.w3.org/feed/docs/rss2.html</docs>
        <generator>awesome</generator>
        <image>
            <title>Feed Title</title>
            <url>http://example.com/image.png</url>
            <link>http://example.com/</link>
        </image>
        <copyright>All rights reserved 2013, John Doe</copyright>
        <atom:link href=\\"https://example.com/atom\\" rel=\\"self\\" type=\\"application/rss+xml\\"/>
    </channel>
</rss>"
`;

exports[`module function rss 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<rss version=\\"2.0\\" xmlns:atom=\\"http://www.w3.org/2005/Atom\\">
    <channel>
        <title>Feed Title</title>
        <link>http://example.com/</link>
        <description>This is my personal feed!</description>
        <lastBuildDate>Fri, 14 Jul 2000 00:00:00 GMT</lastBuildDate>
        <docs>https://validator.w3.org/feed/docs/rss2.html</docs>
        <generator>awesome</generator>
        <image>
            <title>Feed Title</title>
            <url>http://example.com/image.png</url>
            <link>http://example.com/</link>
        </image>
        <copyright>All rights reserved 2013, John Doe</copyright>
        <atom:link href=\\"https://example.com/atom\\" rel=\\"self\\" type=\\"application/rss+xml\\"/>
    </channel>
</rss>"
`;

exports[`module generate simple rss 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<rss version=\\"2.0\\" xmlns:atom=\\"http://www.w3.org/2005/Atom\\">
    <channel>
        <title>Feed Title</title>
        <link>http://example.com/</link>
        <description>This is my personal feed!</description>
        <lastBuildDate>Fri, 14 Jul 2000 00:00:00 GMT</lastBuildDate>
        <docs>https://validator.w3.org/feed/docs/rss2.html</docs>
        <generator>awesome</generator>
        <image>
            <title>Feed Title</title>
            <url>http://example.com/image.png</url>
            <link>http://example.com/</link>
        </image>
        <copyright>All rights reserved 2013, John Doe</copyright>
        <atom:link href=\\"https://example.com/atom\\" rel=\\"self\\" type=\\"application/rss+xml\\"/>
    </channel>
</rss>"
`;

exports[`module multi rss 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<rss version=\\"2.0\\" xmlns:atom=\\"http://www.w3.org/2005/Atom\\">
    <channel>
        <title>Feed Title</title>
        <link>http://example.com/</link>
        <description>This is my personal feed!</description>
        <lastBuildDate>Fri, 14 Jul 2000 00:00:00 GMT</lastBuildDate>
        <docs>https://validator.w3.org/feed/docs/rss2.html</docs>
        <generator>awesome</generator>
        <image>
            <title>Feed Title</title>
            <url>http://example.com/image.png</url>
            <link>http://example.com/</link>
        </image>
        <copyright>All rights reserved 2013, John Doe</copyright>
        <atom:link href=\\"https://example.com/atom\\" rel=\\"self\\" type=\\"application/rss+xml\\"/>
    </channel>
</rss>"
`;

exports[`module multi rss 2`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<rss version=\\"2.0\\" xmlns:atom=\\"http://www.w3.org/2005/Atom\\">
    <channel>
        <title>Feed Title</title>
        <link>http://example.com/</link>
        <description>This is my personal feed!</description>
        <lastBuildDate>Fri, 14 Jul 2000 00:00:00 GMT</lastBuildDate>
        <docs>https://validator.w3.org/feed/docs/rss2.html</docs>
        <generator>awesome</generator>
        <image>
            <title>Feed Title</title>
            <url>http://example.com/image.png</url>
            <link>http://example.com/</link>
        </image>
        <copyright>All rights reserved 2013, John Doe</copyright>
        <atom:link href=\\"https://example.com/atom\\" rel=\\"self\\" type=\\"application/rss+xml\\"/>
    </channel>
</rss>"
`;

exports[`module no type set 1`] = `""`;

exports[`module non-latin rss 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<rss version=\\"2.0\\">
    <channel>
        <title>Популярные новости России и мира</title>
        <link>http://site.ru/feed.xml</link>
        <description>Новости России и мира на сайте site.ru</description>
        <lastBuildDate>Fri, 14 Jul 2000 00:00:00 GMT</lastBuildDate>
        <docs>https://validator.w3.org/feed/docs/rss2.html</docs>
        <generator>https://github.com/nuxt-community/feed-module</generator>
    </channel>
</rss>"
`;

exports[`module object rss 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<rss version=\\"2.0\\" xmlns:atom=\\"http://www.w3.org/2005/Atom\\">
    <channel>
        <title>Feed Title</title>
        <link>http://example.com/</link>
        <description>This is my personal feed!</description>
        <lastBuildDate>Fri, 14 Jul 2000 00:00:00 GMT</lastBuildDate>
        <docs>https://validator.w3.org/feed/docs/rss2.html</docs>
        <generator>awesome</generator>
        <image>
            <title>Feed Title</title>
            <url>http://example.com/image.png</url>
            <link>http://example.com/</link>
        </image>
        <copyright>All rights reserved 2013, John Doe</copyright>
        <atom:link href=\\"https://example.com/atom\\" rel=\\"self\\" type=\\"application/rss+xml\\"/>
    </channel>
</rss>"
`;

exports[`module simple atom 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<feed xmlns=\\"http://www.w3.org/2005/Atom\\">
    <id>http://example.com/</id>
    <title>Feed Title</title>
    <updated>2000-07-14T00:00:00.000Z</updated>
    <generator>awesome</generator>
    <author>
        <name>John Doe</name>
        <email>johndoe@example.com</email>
        <uri>https://example.com/johndoe</uri>
    </author>
    <link rel=\\"alternate\\" href=\\"http://example.com/\\"/>
    <link rel=\\"self\\" href=\\"https://example.com/atom\\"/>
    <subtitle>This is my personal feed!</subtitle>
    <logo>http://example.com/image.png</logo>
    <icon>http://example.com/favicon.ico</icon>
    <rights>All rights reserved 2013, John Doe</rights>
</feed>"
`;

exports[`module simple json 1`] = `
"{
    \\"version\\": \\"https://jsonfeed.org/version/1\\",
    \\"title\\": \\"Feed Title\\",
    \\"home_page_url\\": \\"http://example.com/\\",
    \\"feed_url\\": \\"https://example.com/json\\",
    \\"description\\": \\"This is my personal feed!\\",
    \\"icon\\": \\"http://example.com/image.png\\",
    \\"author\\": {
        \\"name\\": \\"John Doe\\",
        \\"url\\": \\"https://example.com/johndoe\\"
    },
    \\"items\\": []
}"
`;

exports[`module simple rss 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<rss version=\\"2.0\\" xmlns:atom=\\"http://www.w3.org/2005/Atom\\">
    <channel>
        <title>Feed Title</title>
        <link>http://example.com/</link>
        <description>This is my personal feed!</description>
        <lastBuildDate>Fri, 14 Jul 2000 00:00:00 GMT</lastBuildDate>
        <docs>https://validator.w3.org/feed/docs/rss2.html</docs>
        <generator>awesome</generator>
        <image>
            <title>Feed Title</title>
            <url>http://example.com/image.png</url>
            <link>http://example.com/</link>
        </image>
        <copyright>All rights reserved 2013, John Doe</copyright>
        <atom:link href=\\"https://example.com/atom\\" rel=\\"self\\" type=\\"application/rss+xml\\"/>
    </channel>
</rss>"
`;
